### ISSUE TITLE
Optimize mult+SILU in LLama

### ISSUE THREAD

**Issue by cmaryanTT** [tenstorrent/tt-metal#20546]
"Mult+SiLU: 1 tile mult + silu currently takes 4 us" - could use optimization
Tensor details TBD

**Comment by cmaryanTT** [comment]
@KalaivaniMCW - while we wait for the tensor details, can you please set up a unit test with mult + silu with L1 sharded input that results in 32x32 bf16 shards per core, and profile it.

**Comment by johanna-rock-tt** [comment]
BinaryMult+SiLU inside llama model: https://github.com/tenstorrent/tt-metal/blob/b6b39ea797e43142aad6a793d36117f980804f5a/models/demos/llama3_subdevices/tt/llama_mlp.py#L150

**Comment by KalaivaniMCW** [comment]
@johanna-rock-tt 
I see that the dtype is bfloat8_b. Any idea about the input_shapes used.. will they require broadcasting ? 
```
        ff1ff3 = ttnn.mul(
            w1_out_reduced,
            w3_out_reduced,
            input_tensor_a_activations=[ttnn.UnaryOpType.SILU],
            dtype=ttnn.bfloat8_b,
            memory_config=self.model_config["REDUCE_SCATTER_OUT_MEMCFG"],
        )
```

**Comment by KalaivaniMCW** [comment]
> [@KalaivaniMCW](https://github.com/KalaivaniMCW) - while we wait for the tensor details, can you please set up a unit test with Silu(mult(A, B)) with bf16 L1 sharded input that results in 32x32 (1 tile) shards per core, and profile it.
> 
> Testing this with and without fusing (using post-activations) would be good.

@cmaryanTT both with new and old binary ? It seems the model uses pre-activation on `input_a` from the code @johanna-rock-tt shared

**Comment by johanna-rock-tt** [comment]
See details here: https://docs.google.com/spreadsheets/d/1odrhD-hPnn0KnhwWKsv7mztLeBJG_zSXKOFegppiWuo/edit?usp=sharing

**Comment by johanna-rock-tt** [comment]
Interestingly when measured in the demo (including tracing,..) the duration was 4.5 us, in this measurement (simpler unit test, but also running the whole model) it's 2.633 us.

**Comment by johanna-rock-tt** [comment]
I'm investigating where that difference comes from.

**Comment by KalaivaniMCW** [comment]
@cmaryanTT 
I ran a test for this with both bf16 & bf8b dtypes. 
On current main & the commit before Patrick merged binary_ng w legacy flag. We are still defaulting to old binary.
So far, I am looking at the kernel duration metric and I don't see much difference. 

Results: [link](https://docs.google.com/spreadsheets/d/1UQAkwlJ-bCjeSGaxRu1AI8fXoeC-gTQz0Dt8Rci60CY/edit?usp=sharing)

<img width="879" alt="Image" src="https://github.com/user-attachments/assets/be1df304-8a42-42b3-9f30-5ffdfad6d1f8" />

After binary_ng PR

<img width="866" alt="Image" src="https://github.com/user-attachments/assets/b702ca5d-2099-4f1a-a9c7-fe36998f2224" />

**Comment by dchenTT** [comment]
So it is true about 4 us per tile. Is this a sane number? It seems less likely the 10.2 us is real..

**Comment by cmaryanTT** [comment]
Thanks @KalaivaniMCW - next step would be to separate the time spent on SILU from the mult (I suspect it's mostly the former) try a test that disables the SILU (and look at the difference) or runs a SILU in isolation.

Otherwise, the SILU LLK needs some review. That may be something we can ask the LLK team about.

**Comment by cmaryanTT** [comment]
I think the SFPU implementation is here:
https://github.com/tenstorrent/tt-metal/blob/e3cce88789abb9565d9807b82986cea687a03f94/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sigmoid.h#L19
It's just a 5th order polynomial evaluation, which I think would be about 8-10x the cost of the mult.

```
#define POLYVAL5(coef4, coef3, coef2, coef1, coef0, val) \
    ((((coef4 * val + coef3) * val + coef2) * val + coef1) * val + coef0)

inline vFloat sigmoid_piecewise_linear_positive(vFloat val) {
    vFloat result = 0.0f;
    v_if(val >= +5.0f) { result = 1.0f; }
    v_elseif(val > 1.0f && val < 5.0f) {
        result = POLYVAL5(0.00144462f, -0.01055479f, -0.01203685f, 0.24300185f, 0.50437757f, val);
    }
    v_else {
        result = 0.229f * val + 0.5f;  // linear appx as y = 0.229x + 0.5
    }
    v_endif;
    return result;
}
```

**Comment by KalaivaniMCW** [comment]
Hi @johanna-rock-tt @kpaigwar 
cc: @cmaryanTT 

I wanted to confirm something about this test.. are the input tensors created in Interleaved memory config or sharded_config ?

I re-ran the test today with SILU as a separate op and recorded it's duration. 

And I re-ran the mul_silu tests this time with both inputs having the same sharded config. (In my previous test, the default config was used in ttnn.from_torch so inputs were interleaved. The 4 us observed was when only the output tensor is sharded. ) 
These figures are from the latest main tested on Wormhole n150. 

<img width="1042" alt="Image" src="https://github.com/user-attachments/assets/3eb79e02-aab7-45d6-b39d-59aab82a0d16" />

With ttnn.bfloat8_b in old binary i.e use_legacy=True
<img width="1055" alt="Image" src="https://github.com/user-attachments/assets/076c0299-d54b-4ce3-919b-d881a18ff164" />

pytest repro

```
@pytest.mark.parametrize(
    "input_shapes",
    ((torch.Size([1, 1, 32, 960])),),
)
@pytest.mark.parametrize(
    "input_range",
    [
        {"high": 100, "low": -100},
    ],
)
def test_mul_width_sharded(device, input_shapes, input_range):
    high = input_range["high"]
    low = input_range["low"]
    in_data1 = torch.rand((input_shapes), dtype=torch.bfloat16) * (high - low) + low
    in_data2 = torch.rand((input_shapes), dtype=torch.bfloat16) * (high - low + 10) + low
    shard_grid = ttnn.CoreRangeSet(
        {
            ttnn.CoreRange((0, 0), (7, 7)),
        }
    )
    shard_spec = ttnn.ShardSpec(shard_grid, [32, 32], ttnn.ShardOrientation.ROW_MAJOR, ttnn.ShardMode.PHYSICAL)
    input_mem_config = ttnn.MemoryConfig(
        ttnn.types.TensorMemoryLayout.WIDTH_SHARDED, ttnn.types.BufferType.L1, shard_spec
    )

    input_tensor1 = ttnn.from_torch(
        in_data1,
        dtype=ttnn.bfloat16,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        memory_config=input_mem_config,
    )

    input_tensor2 = ttnn.from_torch(
        in_data2,
        dtype=ttnn.bfloat16,
        layout=ttnn.TILE_LAYOUT,
        device=device,
        memory_config=input_mem_config,
    )
    output_tensor = ttnn.multiply(
        input_tensor1,
        input_tensor2,
        memory_config=input_mem_config,
        input_tensor_a_activations=[ttnn.UnaryWithParam(ttnn.UnaryOpType.SILU)],
        use_legacy=True,
    )

    output_tensor = ttnn.to_torch(output_tensor)
    golden_function = ttnn.get_golden_function(ttnn.multiply)
    golden_tensor = golden_function(torch.nn.functional.silu(in_data1), in_data2)
    # golden_tensor = golden_function(in_data1, in_data2)

    pcc, pcc_msg = assert_with_pcc(golden_tensor, output_tensor, 0.999)
    assert pcc
```

**Comment by KalaivaniMCW** [comment]
Silu op - width_sharded config , bfloat8_b, [1, 1, 32, 960]

<img width="980" alt="Image" src="https://github.com/user-attachments/assets/2795d3ad-0118-4ffc-bc42-bf8db7eebdb0" />

**Comment by cmaryanTT** [comment]
Next thing to try is to rewrite this with an FPU LLK. Need to review with LLK team. Also, why not more FPU usage in unary ops?

**Comment by cmaryanTT** [comment]
@nvelickovicTT - can you please comment on feasibility of implementing Silu in the FPU. The polynomial is maybe doable, but the piecewise nature of the full implementation may prevent it on the FPU.

**Comment by johanna-rock-tt** [comment]
Any updates on this issue? Do we know if there's a possible optimisation here? If so, do we haver perf targets?

**Comment by nvelickovicTT** [comment]
@johanna-rock-tt , @cmaryanTT sorry for the delay. We did a bit of a brainstorm what would be possible here.
There is one big elephant in the room, with respect to SiLU FPU implementation, and that is conditional execution.
Conditional execution is not possible on FPU, so the kernel would not be implementable in the same form it is in now on FPU.

A version that we came up with on a back of the envelope is something like this:

```
#define POLYVAL5(coef4, coef3, coef2, coef1, coef0, val) \
    ((((coef4 * val + coef3) * val + coef2) * val + coef1) * val + coef0)

inline vFloat sigmoid_piecewise_linear_positive(vFloat val) {
    vFloat result = 0.0f;
    v_if(val >= +5.0f) { result = 1.0f; } // HANDLE THIS CUTOFF WITH PACKER RELU
    v_elseif(val > 1.0f && val < 5.0f) { // MERGE BELLOW ELSE RANGE INTO THIS ONE
        result = POLYVAL5(0.00144462f, -0.01055479f, -0.01203685f, 0.24300185f, 0.50437757f, val);
    }
    v_else { // MERGE INTO ELSE RANGE ABOVE
        result = 0.229f * val + 0.5f;  // linear appx as y = 0.229x + 0.5
    }
    v_endif;
    return result;
}
```

This would give us a kernel that just does the polynomial part, whose results would be cut-of by packer RELU.
The way this FPU kernel would be implemented is something like this:

```
1. UNPACK coef3 to DEST[0] - unpack 1 row to srcB, elwadd with srcb broadcast to dest (16 cycles/tile)
2. UNPACK coef2 to DEST[1] - same as above, overlap with previous elwadd leads to (8 cycles/tile)
3. UNPACK coef1 to DEST[2] - same as above, overlap with previous elwadd leads to (8 cycles/tile)
4. UNPACK coef0 to DEST[3] - same as above, overlap with previous elwadd leads to (8 cycles/tile)
5. UNPACK one row of coef4 to SRCB, val to SRCA (33 cycles/tile)
6. ELWMUL w/ DEST ACC to DEST[0] - coef4*val + coef3 (8 cycles/tile)
7. MOVD2B from DEST[0] (16 cycles/tile)
8. ELWMUL w/ DEST ACC to DEST[1] (8 cycles/tile)
9. MOVD2B from DEST[1] (16 cycles/tile)
10. ELWMUL w/ DEST ACC to DEST[2] (8 cycles/tile)
11. MOVD2B from DEST[2] (16 cycles/tile)
12. EWLMUL w/ DESTACC to DEST[3] (8 cycles/tile)
```
This is theoretical best estimate, useful only for comparison with SFPU impl and it would be **153 cycles/tile**.
We estimated SFPU kernel to take 32 iterations/tile * (8 cycles for mull/adds in each iteration in polynomial + 2 cycles for load/store in each itreation) = **320 cycles/tile**

Comparing the 2 numbers it seems there could be significant improvement in perf, but also significant changes needed. I am not entirely sure about using packer ReLU in the way intended in this proposal. Definitely worth exploring as a prototype.

**Comment by nvelickovicTT** [comment]
@ttmtrajkovic, @rtawfik01 fyi, take a look at the proposal above and call out if there is something wildly off.

**Comment by amahmudTT** [comment]
> [@ttmtrajkovic](https://github.com/ttmtrajkovic), [@rtawfik01](https://github.com/rtawfik01), [@amahmudTT](https://github.com/amahmudTT) fyi, take a look at the proposal above and call out if there is something wildly off.

What is the fidelity for the ELTWISE Muls ? Asking because there is a bug in ELTWISE Mul that probably makes it not work for  higher precisions.

**Comment by cmaryanTT** [comment]
@rtawfik01 @amahmudTT - can you comment on what the bug is?

**Comment by amahmudTT** [comment]
> [@rtawfik01](https://github.com/rtawfik01) [@amahmudTT](https://github.com/amahmudTT) - can you comment on what the bug is?

With higher fidelity for elementwise multiplication there is a simple bug, that will not result in higher fidelity the fix is from what I know very easy, the harder part is getting a proper test case to test the fix (will require special numbers to make sure the results match)

**Comment by cmaryanTT** [comment]
@nvelickovicTT - I think this is worth trying.

**Comment by johanna-rock-tt** [comment]
Any updates on this? Is someone actively working on it?

**Comment by nvelickovicTT** [comment]
No, no one from LLK is working on this at the moment. It is currently also bellow some other items on the prioritization list from OP team https://github.com/orgs/tenstorrent/projects/107/views/19.
There are 2 possibilities:
- This gets reprioritized higher so that it is soon picked up by someone from LLK team.
- Or someone from the OP team takes a shot at it, and we can help guide through the process of implementing that as LLK.
@cmaryanTT

**Comment by vmilicevicTT** [comment]
Reword for bounty - @nvelickovicTT

**Comment by nvelickovicTT** [comment]
Discussed with @johanna-rock-tt and @KalaivaniMCW offline. SiLU was optimized after this issue was created, which reduced execution time to 2.6us.
It was a stretch if the optimization of moving SiLU to FPU would be worth while from the start, after optimization even less so. Reducing priority to P2.

**Comment by vmilicevicTT** [comment]
@cmaryanTT  to review

**Comment by cmaryanTT** [comment]
Closing - this is not going to be optimized further for now.
